tower/src/buffer/error.rs

```diff
@@ -1,47 +1,12 @@
 //! Error types for the `Buffer` middleware.
 
-use crate::BoxError;
-use std::{fmt, sync::Arc};
-
-/// An error produced by a `Service` wrapped by a `Buffer`
-#[derive(Debug)]
-pub struct ServiceError {
-    inner: Arc<BoxError>,
-}
+use std::fmt;
 
 /// An error produced when the a buffer's worker closes unexpectedly.
 pub struct Closed {
     _p: (),
 }
 
-// ===== impl ServiceError =====
-
-impl ServiceError {
-    pub(crate) fn new(inner: BoxError) -> ServiceError {
-        let inner = Arc::new(inner);
-        ServiceError { inner }
-    }
-
-    // Private to avoid exposing `Clone` trait as part of the public API
-    pub(crate) fn clone(&self) -> ServiceError {
-        ServiceError {
-            inner: self.inner.clone(),
-        }
-    }
-}
-
-impl fmt::Display for ServiceError {
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        write!(fmt, "buffered service failed: {}", self.inner)
```

Comment:
formatting the source error in this errors Display impl and also returning the source instead of your source's source in the Error impl was a bug fwiw

Reply:
FWIW, I think this is because the `Display` impl may have predated `Error::source` being useable...